#70 Role based access control

slashplusweb opened the pull request.

slashplusweb commented on migrations/2020_03_30_101351_create_cms_permissions.php:
```diff
@@ -0,0 +1,51 @@
+            $crud = $parts[1] ?? null;
+            $subGroup = $parts[2] ?? null;
+
+            Permission::create(['name' => $permission, 'group' => $group, 'crud' => $crud, 'sub_group' => $subGroup]);
```
crud are more or less the actions the permissions belong to. like create retreave update delete. actually i didnt find a better name...

first of all I make use of the [Wildcard permissions](https://docs.spatie.be/laravel-permission/v3/basic-usage/wildcard-permissions/). The idea was to allow subcomponents later on, so we have a base group like `posts`, define further actions we want to give permissions to i.e. `view`, and then also allow to refine the groups by allowing named subgroups i.e. a specific field or component within the base group. By using the wildcard `*` all subgroups are allowed.

bambamboole replied:
oh kay, but we have some more permissions than default crud. maybe there should  also permissions for edit the own posts.

I also think  that the name bust be the whole  thing instead  of splitting  this.

slashplusweb replied:
> oh kay, but we have some more permissions than default crud. maybe there should also permissions for edit the own posts.

Mhm i think most changes correspond to the default crud. For example:
- posts.update.* 
_gives the permission to edit every post (this does neither imply any other actions like creating or deleting the complete database row (but editing every detail of the data stored in the row), nor a specific component)_
- posts.update.own-posts
_gives the permission to edit one specific single post (again all data in the db row, but not the row itself)_
- posts.update.all-posts
_gives the permission to edit one specific single post (again all data in the db row, but not the row itself)_

so basically the first two parts of the string, define which group the permission belongs to, and what database action is allowed. the subgroup defines the actual permission (what exactly should be allowed).

But of course, let me know which permission wouldnt fit in this. Maybe I just dont see it at the moment.

> I also think that the name bust be the whole thing instead of splitting this.

at first I felt the same, but after further thinking, I've come to the conclusion that this might be consequent as well as clear (jo can instantly see how strong the permission actually is, just by looking at the 'crud' part). It also allows us to group the permissions into several tables i.e. posts, pages, menus, options.

slashplusweb commented on resources/js/modules/roles/components/PermissionTable.vue:
```diff
@@ -0,0 +1,46 @@
+<template>
+    <table class="table-fixed w-full">
+        <thead>
+        <tr>
+            <th class="w-1/5 px-4 py-2">Permission</th>
+            <th class="w-1/5 px-4 py-2">Create</th>
+            <th class="w-1/5 px-4 py-2">Delete</th>
+            <th class="w-1/5 px-4 py-2">Update</th>
+            <th class="w-1/5 px-4 py-2">View</th>
+        </tr>
+        </thead>
+        <tbody>
+        <PermissionTableRow v-for="(permissionGroup, permissionGroupName) in permissionGroups"
+                            :key="permissionGroupName"
+                            :role="role"
+                            :name="permissionGroupName"
+                            :permissions="permissionGroup"/>
+        </tbody>
+    </table>
```
I considered this, but I came to the conclusion, that this might not be the best idea.

lets discuss this in our channel.

slashplusweb commented on resources/js/modules/roles/components/PermissionTable.vue:
```diff
@@ -0,0 +1,46 @@
+            permissions: Array,
+            name: String,
+        },
+        data() {
+            return {
+                rolePermissions: this.role.permissions,
+            }
+        },
```
jup, still WIP. i will remove it. good catch :)

bambamboole commented on resources/js/components/ResourceForm.vue:
```diff
@@ -55,6 +72,17 @@
             },
+            submitPositions: {
+                type: Array,
+                validator: function (value) {
+                    const valid = ['top', 'bottom'];
+
+                    return value === undefined || difference(value, valid).length === 0;
+                },
```
```suggestion
                validator: (value) => {
                    const valid = ['top', 'bottom'];

                    return value === undefined || difference(value, valid).length === 0;
                },
```

bambamboole commented on resources/js/components/ResourceForm.vue:
```diff
@@ -55,6 +72,17 @@
+
+                    return value === undefined || difference(value, valid).length === 0;
+                },
+                default: function () {
+                    return ['bottom']
+                }
```
```suggestion
                default: () => {
                    return ['bottom']
                }
```

bambamboole commented on resources/js/fields/CheckboxField.vue:
```diff
@@ -0,0 +1,26 @@
+    export default {
+        mixins: [FormField],
+        computed: {
+            inputType() {
+                return this.field.attributes && this.field.attributes.type
+                    ? this.field.attributes.type
+                    : 'checkbox';
+            }
```
this can be removed

bambamboole commented on resources/js/modules/roles/views/Show.vue:
```diff
@@ -0,0 +1,69 @@
+        computed: {
+        },
+        async mounted() {
+            const responseRole = await api(Cms.route('cms.api.resources.show', ['role', this.id]));
+            this.role = responseRole.data.data;
```
```suggestion
            const response = await api(Cms.route('cms.api.resources.show', ['role', this.id]));
            this.role = response.data.data;
```
this is called response everywhere. we should stick to this

bambamboole commented on resources/js/modules/roles/views/Show.vue:
```diff
@@ -0,0 +1,69 @@
+        methods: {
+            handleSuccess(role) {
+                Cms.flash('success', 'Nice one!');
+                // this.$router.push({name: 'role.show', params: {id: post.id}})
```
I think this can be removed

slashplusweb replied:
why?

bambamboole commented on src/Backend/Form/Form.php:
```diff
@@ -112,22 +85,27 @@ protected function afterValidation(array $data): array
     }
 
-    protected function removeNullValues()
+    public function afterCreateResource($resource)
```
```suggestion
    protected function afterCreateResource($resource)
```

this method will be overridden in the form class and is not part of its api :-)

bambamboole commented on src/Backend/Routing/BackendRouter.php:
```diff
@@ -37,6 +39,7 @@ public function registerAuthRoutes()
                 $router->post('/password/forgot', [ForgotPasswordController::class, 'sendResetLinkEmail'])->name('password.email');
                 $router->get('/password/reset/{token}', [ForgotPasswordController::class, 'showNewPassword'])->name('password.reset');
             });
+        $this->router->aliasMiddleware('permission', PermissionMiddleware::class);
```
I dont think we need this on the auth routes.
```suggestion
```

slashplusweb commented:
@bambamboole 
I added better roles with permissions in the seeder and commited your suggestions.
Moreover I added the super-admin role in the login function for tests. but actually i dont know why the mail is not gonna be sent. do you have an idea?

please merge the branch if youre ok with it.
